### PR TITLE
Demo img2img webcam browser

### PR DESCRIPTION
**Pull request by radames**
Demo Notes
* Frontend in Svelte
* Backend: FastAPI / WebSocket / MJPEG stream
* Wrapper is copied and modified here to accept a prompt for `img2img` and `engine_dir`, allowing me to specify the directory and reuse the compiled model in the Docker environment.

All the StreamDiffusion code is on [img2img.py](https://github.com/cumulo-autumn/StreamDiffusion/compare/main...radames:dev/demo-img2img?expand=1#diff-5f7058bed25d1937fa0758334643c19c8d9ae33c43ddc5350a97c4c1c234d04dR67-R88). Please feel free to add any speedup suggestions.
I'm using `t_index_list=[35, 45]`. Is there a way to provide a `strength` on a 0-1 scale?

### CODE REVIEW

**Comment by cumulo-autumn** [comment]
We will make a noise scheduler function and realize parametric operations such as noise strength 0.0~1.0 and num_denoising_steps 1~50. We will also do the review for the PR soon!

File: demo/realtime-img2img/img2img.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            prompt=default_prompt,
+            negative_prompt=default_negative_prompt,
+            num_inference_steps=50,
+            guidance_scale=1.2,
+        )
+
+    def predict(self, params: "Pipeline.InputParams") -> Image.Image:
+        image_tensor = self.stream.preprocess_image(params.image)
+        for _ in range(self.stream.batch_size - 1):
+            self.stream(image=image_tensor, prompt=params.prompt)
```
Comment (GradientSurfer) [inline demo/realtime-img2img/img2img.py:101]: - is repeating inference multiple times needed here? I don't believe so, but if it is I would be keen to know why
- in the `"img2img"` mode it appears `stream.preprocess_image()` is already called, so I believe that is also unnecessary here
- it appears `img2img()` in `utils/wrapper.py` doesn't accept a `prompt` argument so I'm curious to know how the call to `stream()` works correctly here - presumably I've overlooked something because I see the prompt change clearly taking effect in the demo gifs

File: demo/realtime-img2img/main.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                    params = SimpleNamespace(**params.dict())
+                    if info.input_mode == "image":
+                        image_data = await websocket.receive_bytes()
+                        if len(image_data) == 0:
+                            await websocket.send_json({"status": "send_frame"})
+                            continue
+                        params.image = bytes_to_pil(image_data)
+                    await user_data.update_data(user_id, params)
+                    await websocket.send_json({"status": "wait"})
+                    await asyncio.sleep(THROTTLE)
```
Comment (GradientSurfer) [inline demo/realtime-img2img/main.py:107]: minor nit: I believe this `sleep()` is unnecessary, since `await websocket.receive_json()` should block until the client sends a message to the backend server

File: demo/realtime-img2img/main.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+                async def generate():
+                    websocket = user_data.get_websocket(user_id)
+                    last_params = SimpleNamespace()
+                    while True:
+                        last_time = time.time()
+                        params = await user_data.get_latest_data(user_id)
+                        if not vars(params) or params.__dict__ == last_params.__dict__:
+                            await websocket.send_json({"status": "send_frame"})
+                            continue
```
Comment (GradientSurfer) [inline demo/realtime-img2img/main.py:130]: minor nit: might be a good idea to `await asyncio.sleep(THROTTLE)` before these two `continue` statements, to avoid spamming a slow client

**Review by GradientSurfer — COMMENTED** [review]
Looks cool, nice work @radames! I've been tinkering with a strikingly similar set of changes, but using a canvas with drawing tools instead of webcam input. 

If you and the team don't mind unsolicited feedback, I'll leave a review and share a few suggestions/thoughts that I hope you find helpful:

1. **Batch inference** 
It appears image frames are processed one at a time in this demo, but batching multiple frames together for higher throughput (& FPS)  should result in a smoother experience (at the expense of increased latency).

2. **Circular buffer & continuous streaming**
It looks like the server requests the client to send a frame - instead of this request/response cycle, the client could continuously stream image frames to the server which would maintain a circular buffer that can then be used to perform batch inference. Notably `examples/screen/main.py` uses that approach.

3. **No separate endpoint/stream for returning generated images**
The generated image could be returned to the client via the websocket connection, instead of via a separate API endpoint. This could be a minor code simplification, and notably would sidestep the linked chromium bug (so we could avoid sending frames twice to every browser that isn't firefox).

4. **Return raw pixels in RGBA format**
Generated images can be returned to the client in raw RGBA pixel format and then directly written to the canvas. This may be a relatively minor optimization, but it avoids the overhead of transforming to & from JPEG format and any associated lossy compression.

5. **Integrate `wrapper.py` modifications**
The modifications to accept a prompt for img2img and to support `engine_dir` are great and seem well contained, those ought to be integrated into the canonical `wrapper.py` so there is no unnecessary duplication of code or maintenance burden.

Perhaps these ideas could be addressed here or in future PRs (or not at all), either way I'd be happy to discuss or collaborate further on details - feel free to reach out.

File: demo/realtime-img2img/img2img.py (hunk 1 above)
Comment (radames) [reply demo/realtime-img2img/img2img.py:101]: I copy the batches call from here, it might be necessary, I'm no sure, wdyt @cumulo-autumn, same image tensor called `batch_size - 1` times 

https://github.com/cumulo-autumn/StreamDiffusion/blob/62596fd39b86a3088e28b2f2c58caa36bc069c50/examples/img2img/single.py#L97-L103

do you mean the `self.stream(image=image_tensor` can be a PIL image? my guess is I need to transform the `params.image` to a tensor every single predict `image_tensor = self.stream.preprocess_image(params.image)`

**Comment by radames** [comment]
hi @GradientSurfer, thanks for detailed response! I really appreciate the feedback, I'm happy to address some of your points on this PR, also if you're interested in collaborating, please send edits, commits. do you have PR edit access?

> Looks cool, nice work @radames! I've been tinkering with a strikingly similar set of changes, but using a canvas with drawing tools instead of webcam input.
> 
> If you and the team don't mind unsolicited feedback, I'll leave a review and share a few suggestions/thoughts that I hope you find helpful:
> 
> 1. **Batch inference**
>    It appears image frames are processed one at a time in this demo, but batching multiple frames together for higher throughput (& FPS)  should result in a smoother experience (at the expense of increased latency). 

Addressing number 2 here, we can try a batching approach!

> 2. **Circular buffer & continuous streaming**
>    It looks like the server requests the client to send a frame - instead of this request/response cycle, the client could continuously stream image frames to the server which would maintain a circular buffer that can then be used to perform batch inference. Notably `examples/screen/main.py` uses that approach.

Ohh yes that makes a lot of sense, In my original demo with LCM, I did use an async `queue()`, but back in time, the inference was slow, and the result was a lagged video, thus I decided to switch to ping/pong approach.

> 3. **No separate endpoint/stream for returning generated images**
>    The generated image could be returned to the client via the websocket connection, instead of via a separate API endpoint. This could be a minor code simplification, and notably would sidestep the linked chromium bug (so we could avoid sending frames twice to every browser that isn't firefox).
Yes, that's a great point, MJPEG stream seems a bit awkward, and buggy on Chrome, Ideally it would be on WebRTC, but I was looking into performance and simplicity. An open socket jpeg streaming looked faster to me compared to sending blobs over websocket, it needs an extra decoding processing to send the bytes to the ``.  However, this demo seems very fast and it's doing the blob over websockets -> `` https://www.fal.ai/camera

> 4. **Return raw pixels in RGBA format**
>    Generated images can be returned to the client in raw RGBA pixel format and then directly written to the canvas. This may be a relatively minor optimization, but it avoids the overhead of transforming to & from JPEG format and any associated lossy compression.

Yes you're right, however I did the canvas to normalize the webcam image, cropping on the desired size, this could be done on the backend, whichever is faster.

> 5. **Integrate `wrapper.py` modifications**
>    The modifications to accept a prompt for img2img and to support `engine_dir` are great and seem well contained, those ought to be integrated into the canonical `wrapper.py` so there is no unnecessary duplication of code or maintenance burden.
Done on PR #66 , and when it's merged I can update it here.

> 
> Perhaps these ideas could be addressed here or in future PRs (or not at all), either way I'd be happy to discuss or collaborate further on details - feel free to reach out.

**Comment by GradientSurfer** [comment]
@radames I do not have PR edit access here, @cumulo-autumn perhaps you would consider granting collaborator access?

File: demo/realtime-img2img/img2img.py (hunk 1 above)
Comment (cumulo-autumn) [reply demo/realtime-img2img/img2img.py:108]: Lines 100 and 101 of img2img.py are unnecessary.

We still need to do inference multiple times in img2img/single.py demo because we need to push out the initial latent tensor, which is initialized with torch.zeros(). 
```py
for _ in range(stream.batch_size - 1): 
	     stream(image=image_tensor) 
```
Since we are using Stream Batch, which batching the denoising steps, we need to push out initial n-1 images made from torch.zeros tensor first and then get the actual img2img result after that, where n is the number of denoising steps (n = len(t_index_list) ).

The concepts of Stream Batch are shown in Figure 2. and Figure 5. of our paper. https://arxiv.org/abs/2312.12491

If the Stream Batch is filled with zero tensors at the start, you need to call `stream()` multiple times to discard them (in this case, it's not necessary to use the same image multiple times). On the other hand, once the batch is filled with image tensors in the process of denoising, there's no need to discard those images in the middle of the StreamBatch. Instead, you can simply call stream() once within the while loop.

Comment (cumulo-autumn) [reply demo/realtime-img2img/img2img.py:108]: So the short answers are as follows,
realtime-img2img/img2img.py ... repeatedly calling `stream()` is unnecessary.

img2img/single.py ... repeatedly calling `stream()` is necesary.

Comment (radames) [reply demo/realtime-img2img/img2img.py:108]: perfect, removed unnecessary calls

**Comment by radames** [comment]
hi @cumulo-autumn , I think it's good now. I've fixed a couple of uncaught exceptions. One important note, while the server and the client were designed to accept multiple queued connections, the wrapper and `StreamDiffusionWrapper` are not working well in that regard, i.e. the buffer across `stream.stream` is shared, so if you open multiple browser tabs and switch the prompt and webcams, you'll notice the images are leaking across tabs. For instance when using diffusers `pipe(...)` it's possible to have queue calls, as long as they're quick inference, example [here](https://huggingface.co/spaces/radames/Real-Time-SD-Turbo). 
ps. please pull and test again if you can on Windows

**Comment by cumulo-autumn** [comment]
> @radames I do not have PR edit access here, @cumulo-autumn perhaps you would consider granting collaborator access?

Hi @GradientSurfer . Thank you for your valuable PR submissions in the past, and for your many meaningful suggestions this time as well! Regarding PR edit access, currently, we are keeping it within a group of acquaintances, so please allow us to hold off on adding new PR edit access for now. However, we are very much open to more discussions and PRs in the future, so we definitely want to continue doing those! (I apologize for the late response this time, as it has been a busy end-of-year period. Also, I really appreciate your prompt and valuable feedback on this PR.) We will consider our policy on adding new PR edit access in the future!

**Review by cumulo-autumn — APPROVED** [review]

**Comment by cumulo-autumn** [comment]
> hi @cumulo-autumn , I think it's good now. I've fixed a couple of uncaught exceptions. One important note, while the server and the client were designed to accept multiple queued connections, the wrapper and `StreamDiffusionWrapper` are not working well in that regard, i.e. the buffer across `stream.stream` is shared, so if you open multiple browser tabs and switch the prompt and webcams, you'll notice the images are leaking across tabs. For instance when using diffusers `pipe(...)` it's possible to have queue calls, as long as they're quick inference, example [here](https://huggingface.co/spaces/radames/Real-Time-SD-Turbo). ps. please pull and test again if you can on Windows

Hi @radames. Thank you for the update! It works perfectly in my environment too! I am going to merge it.

**Comment by openSourcerer9000** [comment]
@cumulo-autumn I'm not seeing denoising_strength param in the latest repo. How do we set it? Thanks,
